### ISSUE TITLE
Make the Document binary serialization format more compact.

### ISSUE THREAD

**Issue by fulmicoton** [quickwit-oss/tantivy#903]
The DocStore builds blocks of contiguous document serialized in an adhoc simple binary format.
These blocks are then compressed.

The format goes:
- field: u32
- type tag: u8
- value: specific to the type but for instance u64 simply take 8 byts.

@ppodolsky noticed that despite the compression, it might be possible to shave off a few percent of storage by changing the encoding. This might be especially useful when running 

Field, unsigned int, and date could be using variable byte encoding.
Signed int could use zigzag encoding.

This issue is open to new contributor. Variable int encoding is available (search `VInt`). Zigzag is not implemented yet.
Please add a bench to ensure there is not a catastrophic change in CPU usage when using LZ4, and please try to quantify the size impact on a pathological schema (100 uint for instance).

**Comment by ppodolsky** [comment]
Numbers of the day: using of the varint instead of u32 in field_id decreases brotli-compressed file-size exactly for 1%

BTW, the question bothered me all day (especially taking into account that you are from Google) - why does your varint use most significant bit set for marking the last byte and google's protobuf varint does the opposite by unsetting the msb of the last byte?

**Comment by fabricedesre** [comment]
Is there a reason to not use bincode since it uses varint and zigzag encoding?

**Comment by Divyesh-k** [comment]
It switches the doc store's field ids and integers to variable-length encoding (VInt, plus zig-zag for signed integers), behind a new V3 format version so existing segments still read via the old fixed-width path — no reindexing needed.

On a numeric-heavy schema (20k docs × 100 u64 fields + 1 i64), the LZ4-compressed store shrank from 12,714,578 → 9,445,506 bytes (~25.7%). Dates were intentionally left fixed-width, since VInt would grow modern nanosecond timestamps. Includes zig-zag roundtrip and pathological-schema store tests.

**Comment by fulmicoton-dd** [comment]
@Divyesh-k rollling out a forward change like this one in distributed systems will require upgrading all nodes to allow reading v3 and v2, but keeping writing v2, and then in a second step enabling writing V3.

Currently this is not possible in tantivy, and the current change is not worth the little dance anyway, so I suggest we revisit and cut a v3 only after change have accumulated enough to make the migration effort worthy enough.

**Comment by Divyesh-k** [comment]
Makes total sense — a standalone V3 bump isn't worth the migration dance just for the size win, and I agree it's better to batch it with other format changes so one migration carries several improvements at once.

Two things from my side:

- I'm happy to park this branch and keep it ready so the encoding work can be folded into the next V3 batch whenever enough format changes have accumulated to justify the migration. Just let me know if you'd prefer I leave it as a draft PR or simply track it here on the issue.

- If it's useful, I'd also be glad to take a first crack at the read-new / write-old compatibility mechanism you mentioned is currently missing — i.e. letting nodes read a newer format version while still writing the old one, so future V3 rollouts (this one included) can be staged safely across a cluster. That seems like the actual blocker for any forward format change, so it might make sense to land that groundwork first.

**Comment by Divyesh-k** [comment]
@fulmicoton-dd can you please check and let me know
